### PR TITLE
Init

### PR DESCRIPTION
**Pull request by delvedor**
First implementation.
Heavily inspired by [`zipkin-instrumentation-*`](https://github.com/openzipkin/zipkin-js/tree/master/packages).

Feedbacks?
cc @lucamaraschi

### CODE REVIEW

**Comment by marcbachmann** [comment]
You might want to directly use opentracing instead of zipkin.
Or are there any disadvantages there that I don't know?

**Comment by delvedor** [comment]
@marcbachmann actually a nice idea could be a generic fastify tracing, where based on the options the user decides which protocol use.

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+language: node_js
+
+node_js:
+  - "9"
+  - "8"
+  - "6"
+  - "4"
```
Comment (mcollina) [inline .travis.yml]: I would _just_ support 8+ for this, and leverage async_hooks as much as possible. I'm not a fan of https://github.com/othiym23/node-continuation-local-storage. I'm not 100% sure how this is compatible with the Zipkin module. See https://www.npmjs.com/package/cls-hooked.

File: README.md [hunk 2]
```diff
@@ -1 +1,47 @@
...
+const recorder = new BatchRecorder({
+  logger: new HttpLogger({
+    endpoint: `${zipkinBaseUrl}/api/v2/spans`,
+    jsonEncoder: JSON_V2
+  })
+})
+
+const ctxImpl = new CLSContext('zipkin')
+const localServiceName = 'service-name'
+const tracer = new Tracer({ ctxImpl, recorder, localServiceName })
```
Comment (mcollina) [inline README.md]: I would prefer if fastify-zipkin abstracts away most of this complexity.

File: index.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    port: opts.port || 0
+  })
+
+  try {
+    // we must register a custom error handler
+    // otherwise all the 404 requests
+    // will not be catched by our hooks
+    fastify.setNotFoundHandler(basic404)
+  } catch (err) {
+    // a custom error handler is already present
```
Comment (mcollina) [inline index.js]: we should check if that is indeed the case, and rebounce the other exceptions.

File: index.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  fastify.addHook('onResponse', onResponse)
+
+  function onRequest (req, res, next) {
+    tracer.scoped(() => {
+      var id = instrumentation.recordRequest(
+        req.method,
+        url.format(req.url),
+        readHeader.bind(req.headers)
+      )
+      res._zipkinId = id
```
Comment (mcollina) [inline index.js]: As this is using CLS, do we _really_ need to attach this id to the response?

File: index.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+  fastify.addHook('onRequest', onRequest)
+  fastify.addHook('onResponse', onResponse)
+
+  function onRequest (req, res, next) {
+    tracer.scoped(() => {
+      var id = instrumentation.recordRequest(
+        req.method,
+        url.format(req.url),
+        readHeader.bind(req.headers)
```
Comment (mcollina) [inline index.js]: bind would be very slow on Node 4 and 6.

File: test.js [hunk 6]
```diff
@@ -0,0 +1,157 @@
+'use strict'
+
+const test = require('tap').test
+const sinon = require('sinon')
```
Comment (mcollina) [inline test.js]: I'm not a fan ok mocking. I would really prefer to have at least 1 test against the live system.
However, I'm ok with the change.

File: README.md (hunk 2 above)
Comment (lucamaraschi) [reply README.md]: I agree

**Comment by lucamaraschi** [comment]
@delvedor shall we just open an issue and continue the discussion about `fastify-tracing` over there?I personally think it's a very interesting path to analyze and consider.

**Comment by delvedor** [comment]
@lucamaraschi please go ahead!
Write down the idea and we can continue to discuss over there :)
Regarding this repo I think we can finish to develop it, and in case merge it with a future general purpose tracing.

**Comment by thearegee** [comment]
@mcollina - does this look like its going to be added?

**Comment by mcollina** [comment]
I think the implementation of this has stalled.

@delvedor @lucamaraschi, would you like to bring this over the finishing line? @thearegee would you like to champion this?

**Comment by thearegee** [comment]
I don't think I'm enough of an expert on this topic to champion this, however I agree with @marcbachmann that it should be more abstract and use OpenTracing directly.

I know there is a lot of traction with [jaeger](https://github.com/jaegertracing/jaeger), so maybe the plug in being less directly tied to zipkin might make more sense. However maybe this is being discussed on another repo?

**Comment by Eomm** [comment]
Could we list what should be done to this version in order to release it?

**Pull request merged**